### PR TITLE
Fixes and AMSoftmax support in pytorchcv wrapper

### PR DESCRIPTION
**Pull request by sovrasov**
@evgeny-izutov please take a look at the changes in metric learning losses. Seems like now pytorch disables making optimizer.step() between two related backwards, so I had to move the call of end_iteration().

### CODE REVIEW

File: torchreid/losses/metric.py [hunk 1]
```diff
@@ -301,7 +301,7 @@ def init_iteration(self):
     def end_iteration(self):
         """Finalizes a training iteration"""
 
-        self.last_loss_value.backward(retain_graph=True)
+        self.last_loss_value.backward()
```
Comment (evgeny-izutov) [inline torchreid/losses/metric.py]: It looks like the gradients have been already calculated during the stage: `total_loss.backward(retain_graph=self.enable_metric_losses)`.

File: torchreid/engine/image/am_softmax.py [hunk 2]
```diff
@@ -252,7 +257,7 @@ def _single_model_losses(self, model, train_records, imgs, obj_ids, n_iter, mode
 
                 ml_loss_module.init_iteration()
                 ml_loss, ml_loss_summary = ml_loss_module(embd, trg_logits, trg_obj_ids, n_iter)
-                ml_loss_module.end_iteration()
+                #ml_loss_module.end_iteration()
```
Comment (evgeny-izutov) [inline torchreid/engine/image/am_softmax.py]: In this way, please remove this line.

File: torchreid/losses/metric.py (hunk 1 above)
Comment (sovrasov) [reply torchreid/losses/metric.py]: Made this optional it.

File: torchreid/engine/image/am_softmax.py (hunk 2 above)
Comment (sovrasov) [reply torchreid/engine/image/am_softmax.py]: Removed

File: torchreid/engine/image/am_softmax.py [hunk 3]
```diff
@@ -206,13 +206,13 @@ def forward_backward(self, data):
 
             total_loss += mutual_loss / float(num_mutual_losses)
 
-        total_loss.backward(retain_graph=self.enable_metric_losses)
+        total_loss.backward()
```
Comment (evgeny-izutov) [inline torchreid/engine/image/am_softmax.py:209]: I'm not sure that it will work without `retain_graph` (for ml losses) - as far as I know each call of `step()` method clears grads and any future call of `step()` will fail. Maybe I'm wrong.

Comment (sovrasov) [reply torchreid/engine/image/am_softmax.py:209]: According to the documentation of pytorch 1.7 the step method doesn't affect .grad fields
```
Unless otherwise specified, this function should not modify the .grad field of the parameters.
```
https://pytorch.org/docs/stable/optim.html#torch.optim.Optimizer.step

Comment (sovrasov) [reply torchreid/engine/image/am_softmax.py:209]: Originally retain_graph was enabled to allow the second backward.

Comment (evgeny-izutov) [reply torchreid/engine/image/am_softmax.py:209]: Did you have a chance to run it just for single iteration?

Comment (sovrasov) [reply torchreid/engine/image/am_softmax.py:209]: Yes, I performed 500 iterations on imagenet with the center and center push losses enabled

Comment (evgeny-izutov) [reply torchreid/engine/image/am_softmax.py:209]: Ok

**Review by evgeny-izutov — APPROVED** [review]
